sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -178,6 +179,27 @@ class Analyzer(
     }
   }
 
+  /**
+   * Replaces queries of the form "SELECT expr FROM A GROUP BY 1"
+   * with a query of the form "SELECT expr FROM A GROUP BY expr"
+   */
+  object ResolveGroupByReferences extends Rule[LogicalPlan] {
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case Aggregate(groups, aggs, child) =>
+        Aggregate(groups.map(group => group match {
```

Comment:
Nit: would you mind assigning this to an intermediate variable.  this is a little hard to parse visually.